### PR TITLE
fix: ignoring jetpack ability when player is in GameMode

### PR DESCRIPTION
**Pull request by DevSrSouza**
Hi folks! Small contribution for ignoring Removable of Fly when the player is with Jetpack and in GameMode Creative or Spectator.

This also disable losing chargeable or even requiring to have any energy to show particles and sound. 

There is a bunch of if statements, I know, but this was to keep the "Jetpack behavior and effects" while in Creative.

### CODE REVIEW

File: src/main/kotlin/xyz/xenondevs/nova/jetpacks/ability/JetpackFlyAbility.kt [hunk 1]
```diff
@@ -51,14 +62,23 @@ class JetpackFlyAbility(player: Player, flySpeed: Provider<Float>, energyPerTick
...
+        } else {
+            player.flySpeed = previousFlySpeed
+        }
         
         val chargeable = novaItem.getBehavior(Chargeable::class)!!
         val energyLeft = chargeable.getEnergy(jetpackItem)
         overlay.percentage = energyLeft / chargeable.options.maxEnergy.toDouble()
         
-        if (energyLeft > energyPerTick) {
+        if (energyLeft > energyPerTick || isValidGameMode().not()) {
```
Comment (NichtStudioCode) [inline src/main/kotlin/xyz/xenondevs/nova/jetpacks/ability/JetpackFlyAbility.kt]: prefer the ! operator

File: src/main/kotlin/xyz/xenondevs/nova/jetpacks/registry/Abilities.kt [hunk 2]
```diff
@@ -8,6 +9,10 @@ import xyz.xenondevs.nova.jetpacks.Jetpacks
 import xyz.xenondevs.nova.jetpacks.ability.JetpackFlyAbility
 import xyz.xenondevs.nova.util.data.getFloat
 
+private val IGNORED_GAME_MODES = configReloadable {
+    val modesRaw = NovaConfig["jetpacks:config"].getStringList("ignored_game_mode")
+    GameMode.values().filter { gameMode -> modesRaw.any { it.equals(gameMode.name, ignoreCase = true) } }
+}
```
Comment (NichtStudioCode) [inline src/main/kotlin/xyz/xenondevs/nova/jetpacks/registry/Abilities.kt]: This should probably be in JetpackFlyAbility if it's the same across all jetpack types

File: src/main/resources/configs/config.yml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+ignored_game_mode:
```
Comment (NichtStudioCode) [inline src/main/resources/configs/config.yml]: I think this should be `ignored_game_modes`? It's also plural in code.

**Review by NichtStudioCode — CHANGES_REQUESTED** [review]
Please also reformat your code with the [xenondevs codestyle](https://github.com/xenondevs/Nova/blob/main/codestyle.xml)

**Comment by DevSrSouza** [comment]
> Please also reformat your code with the [xenondevs codestyle](https://github.com/xenondevs/Nova/blob/main/codestyle.xml)

I was already using it, the IntelliJ format does not change anyway on the current source code

File: src/main/kotlin/xyz/xenondevs/nova/jetpacks/ability/JetpackFlyAbility.kt [hunk 4]
```diff
@@ -17,48 +20,65 @@ import xyz.xenondevs.nova.util.item.novaItem
...
+    private val ignoredGameModes: List<GameMode> by configReloadable {
+        val modesRaw = NovaConfig["jetpacks:config"].getStringList("ignored_game_modes")
+        GameMode.values().filter { gameMode -> modesRaw.any { it.equals(gameMode.name, ignoreCase = true) } }
+    }
+
     private val wasFlying = player.isFlying
     private val wasAllowFlight = player.allowFlight
     private val previousFlySpeed = player.flySpeed
-    
+
```
Comment (NichtStudioCode) [inline src/main/kotlin/xyz/xenondevs/nova/jetpacks/ability/JetpackFlyAbility.kt]: The linked codestyle.xml enables "keep indents on empty lines"

File: src/main/kotlin/xyz/xenondevs/nova/jetpacks/ability/JetpackFlyAbility.kt [hunk 5]
```diff
@@ -17,16 +19,14 @@ import xyz.xenondevs.nova.util.broadcast
...
-    
+class JetpackFlyAbility(player: Player, flySpeed: Provider<Float>, energyPerTick: Provider<Long>) : Ability(player) {
+
     private val flySpeed: Float by flySpeed
     private val energyPerTick: Long by energyPerTick
-    private val ignoredGameModes: List<GameMode> by ignoredGameModes
+    private val ignoredGameModes: List<GameMode> by configReloadable {
+        val modesRaw = NovaConfig["jetpacks:config"].getStringList("ignored_game_modes")
+        GameMode.values().filter { gameMode -> modesRaw.any { it.equals(gameMode.name, ignoreCase = true) } }
+    }
```
Comment (NichtStudioCode) [inline src/main/kotlin/xyz/xenondevs/nova/jetpacks/ability/JetpackFlyAbility.kt]: That's not what I meant. It should still a top level declaration.

**Comment by NichtStudioCode** [comment]
Thanks, I'll fix the remaining issues myself.

**Pull request merged**